Use mach_vm_map and copy instead of mach_vm_remap in setPermissionsOfConfigPage

ddegazio opened the pull request:
#### 4e67c38f20d95e6f6804d0b5242e37934cc3564a
<pre>
Use mach_vm_map and copy instead of mach_vm_remap in setPermissionsOfConfigPage
<a href="https://bugs.webkit.org/show_bug.cgi?id=290878">https://bugs.webkit.org/show_bug.cgi?id=290878</a>
<a href="https://rdar.apple.com/148277754">rdar://148277754</a>

Reviewed by Mark Lam.

Reverts earlier change adopting mach_vm_remap in WTF Config initialization,
we still need to preserve any pre-WTF elements of g_config but instead of
using mach_vm_remap, we now just copy the earlier elements to a side buffer
and copy them back in once we&apos;ve set the permissions. This avoids depending
on mach_vm_remap on platforms that did not previously use the syscall (namely
non-JIT Apple platforms).

* Source/WTF/wtf/WTFConfig.cpp:
(WTF::setPermissionsOfConfigPage):

Canonical link: <a href="https://commits.webkit.org/293128@main">https://commits.webkit.org/293128@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b32465835a6cbd59ff55bef8ae9c8a1f50f7987a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97833 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/17458 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/7677 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102945 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/48362 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17752 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25911 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74497 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31672 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100836 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13443 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/88418 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54854 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13220 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6343 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/47805 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/90509 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/83210 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6416 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/105325 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/96456 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24913 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/18144 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/83518 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/25285 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84591 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82954 "Passed tests") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20970 "Build is in progress. Recent messages:") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27561 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5244 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/18520 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24874 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/30043 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/120081 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/24696 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/33680 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/28010 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/26270 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

MenloDorian reviewed (approved):
r=me

kmiller68 commented on Source/WTF/wtf/WTFConfig.cpp:
```diff
@@ -73,13 +73,19 @@ void setPermissionsOfConfigPage()
-        auto prot = VM_PROT_READ | VM_PROT_WRITE;
 
         auto attemptVMMapping = [&] {
-            // We use remap instead of map here to avoid zeroing the config,
-            // since WebKit components like bmalloc can potentially use the
-            // config prior to WTF initialization.
-            return mach_vm_remap(mach_task_self(), &addr, ConfigSizeToProtect, pageSize() - 1, flags, mach_task_self(), addr, false, &prot, &prot, VM_INHERIT_DEFAULT);
```
Is there a reason to not try remap first then fallback to the new code?

kmiller68 replied:
That should be fine even if they don't have remap in the sandbox.

MenloDorian replied:
That's possible, but now it makes the code work 2 way and makes it fallout only happen on one platform, and therefore is harder to test for.  It's not a big deal, so I'm ok with either solution.

ddegazio replied:
I think I agree with Mark here, it keeps the code overall simpler and avoids depending on `mach_vm_remap` at all on platforms where we didn't before. And I don't think we really lose much using the copying solution.

ddegazio committed:
Use mach_vm_map and copy instead of mach_vm_remap in setPermissionsOfConfigPage
https://bugs.webkit.org/show_bug.cgi?id=290878
rdar://148277754

Reviewed by Mark Lam.

Reverts earlier change adopting mach_vm_remap in WTF Config initialization,
we still need to preserve any pre-WTF elements of g_config but instead of
using mach_vm_remap, we now just copy the earlier elements to a side buffer
and copy them back in once we've set the permissions. This avoids depending
on mach_vm_remap on platforms that did not previously use the syscall (namely
non-JIT Apple platforms).

* Source/WTF/wtf/WTFConfig.cpp:
(WTF::setPermissionsOfConfigPage):

Canonical link: https://commits.webkit.org/293128@main
